File: backend/src/deaths_door/routes/sounds.py
from fastapi import APIRouter, Path
from fastapi.exceptions import HTTPException
from pydantic import BaseModel, Field

from ..sound_fx import SoundFX, SoundName, sounds

router = APIRouter(prefix="/sounds", tags=["Sounds"])


class PlaySoundResponse(BaseModel):
    """Response after playing a sound effect."""

    status: str = Field(..., description="Operation status", examples=["success"])
    sound: str = Field(..., description="Name of the sound that was played", examples=["bell"])


@router.get(
    "/play/{name}",
    responses={
        404: {"description": "Sound not found"},
        500: {"description": "Failed to play sound"},
    },
)
async def play_sound(
    name: str = Path(..., description="Name of the sound effect to play", examples=["bell"])
) -> PlaySoundResponse:
    """Play a sound effect."""
    sound_name = SoundName.from_str(name)

    if sound_name is None:
        raise HTTPException(status_code=404, detail="Sound not found")

    try:
        SoundFX().play(sound_name)
        return PlaySoundResponse(status="success", sound=name)
    except Exception as e:
        raise HTTPException(status_code=500, detail="Failed to play sound") from e


@router.get("/list")
async def list_sounds() -> dict[str, list[str]]:
    """Return the available sounds organized by category."""
    # Convert SoundName enums to strings for JSON serialization
    return {
        category: [sound.value for sound in sound_list]
        for category, sound_list in sounds.items()
    }

File: backend/src/routes/sounds.rs
//! Routes for playing sound effects.

use axum::extract::Path;
use axum::Json;
use serde::Serialize;
use serde_json::{Map, Value};
use utoipa::ToSchema;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;

use crate::app::{AppError, AppResult, AppState};
use crate::sound::{sounds_by_category, SoundFx, SoundName};

pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(play_sound))
        .routes(routes!(list_sounds))
}

#[derive(Serialize, ToSchema)]
pub struct PlaySoundResponse {
    pub status: String,
    pub sound: String,
}

/// Play a sound effect by name.
#[utoipa::path(
    get, path = "/sounds/play/{name}", tag = "Sounds",
    params(("name" = String, Path, description = "Sound name, e.g. death/rooster/drumroll")),
    responses(
        (status = 200, description = "Sound played", body = PlaySoundResponse),
        (status = 404, description = "sound not found"),
        (status = 500, description = "playback failed")
    )
)]
async fn play_sound(Path(name): Path<String>) -> AppResult<Json<PlaySoundResponse>> {
    let sound_name =
        SoundName::from_str(&name).ok_or_else(|| AppError::not_found("Sound not found"))?;
    // `play` blocks until playback starts (or fails to), so run it off the async
    // worker. A decode/device/file failure now surfaces as a 500 with the reason.
    tokio::task::spawn_blocking(move || SoundFx::new().play(sound_name))
        .await
        .map_err(|e| AppError::internal(format!("sound task failed: {e}")))?
        .map_err(|e| AppError::internal(format!("Failed to play '{name}': {e}")))?;
    Ok(Json(PlaySoundResponse {
        status: "success".to_string(),
        sound: name,
    }))
}

/// List all available sounds grouped by category.
#[utoipa::path(
    get, path = "/sounds/list", tag = "Sounds",
    responses((status = 200, description = "Available sounds grouped by category"))
)]
async fn list_sounds() -> Json<Value> {
    // Preserve the Python category ordering.
    let mut map = Map::new();
    for (category, sound_list) in sounds_by_category() {
        let values: Vec<Value> = sound_list
            .iter()
            .map(|s| Value::String(s.value().to_string()))
            .collect();
        map.insert(category.to_string(), Value::Array(values));
    }
    Json(Value::Object(map))
}
